controller/array_action/array_mediator_svc.py

```diff
@@ -512,19 +512,29 @@ def _stop_fcmap(self, fcmap_id):
             if not is_warning_message(ex.my_message):
                 logger.warning("Failed to stop fcmap '{0}': {1}".format(fcmap_id, ex))
 
-    def _stop_and_delete_fcmap(self, fcmap_id):
-        self._stop_fcmap(fcmap_id)
-        self._delete_fcmap(fcmap_id, force=True)
+    def _safe_stop_and_delete_fcmap(self, fcmap):
+        if not self._is_in_remote_copy_relationship(fcmap):
+            self._stop_fcmap(fcmap.id)
+            self._delete_fcmap(fcmap.id, force=True)
 
     def _safe_delete_fcmaps(self, object_name, fcmaps):
-        unfinished_fcmaps = [fcmap for fcmap in fcmaps
-                             if fcmap.status != FCMAP_STATUS_DONE or fcmap.copy_rate == "0"]
-        if unfinished_fcmaps:
-            raise array_errors.ObjectIsStillInUseError(id_or_name=object_name,
-                                                       used_by=unfinished_fcmaps)
+        fcmaps_to_delete = []
+        fcmaps_in_use = []
+
         for fcmap in fcmaps:
+            if not self._is_in_remote_copy_relationship(fcmap):
+                if fcmap.status != FCMAP_STATUS_DONE or fcmap.copy_rate == "0":
+                    fcmaps_in_use.append(fcmap)
+                else:
+                    fcmaps_to_delete.append(fcmap)
+        if fcmaps_in_use:
+            raise array_errors.ObjectIsStillInUseError(id_or_name=object_name, used_by=fcmaps_in_use)
+        for fcmap in fcmaps_to_delete:
             self._delete_fcmap(fcmap.id, force=False)
 
+    def _is_in_remote_copy_relationship(self, fcmap):
+        return fcmap.rc_controlled == YES
+
     def _delete_object(self, cli_object, is_snapshot=False):
         object_name = cli_object.name
         fcmap_as_target = self._get_fcmap_as_target_if_exists(object_name)
```

Comment:
could there be more than 1? now that we support both volume from snapshot and hyperswap?

Reply:
At the moment we use the `_get_fcmap_as_target_if_exists` function in five scenarios:
1. Create volume
2. Create snapshot
3. Delete volume
4. Delete snapshot
5. Expand volume

Can't a volume have more than one snapshot? Why now it's different than the other cases?
Also, when do we expect to have more than one fcmap as target?

Reply:
I think this will answer your 3 questions:
before hyperswap, a volume (that we created) could not have been a target in more than 1 fcmap, because that would cause data corruption/loss.
e.g. for
```
v1 --fcmap1-> s1 --fcmap2-->v2
 |
 `--fcmap3--> s2
 ```
 the fcmaps would have:
 ```
fcmap1.source = v1
fcmap1.target = s1
 
fcmap2.source = s1
fcmap2.target = v2
 
fcmap3.source = v1
fcmap3.target = s2
 ```
notice that no volume/snapshot has more than 1 fcmap "as input".


but now with hyperswap, I'm not sure if that is still the case.
@ArbelNathan ?

Reply:
Thanks for the explanation.
In the [code](https://github.com/IBM/ibm-block-csi-driver/blob/a5bba0bf27b34a381f9ea1a538f1407ba6df0370/controller/array_action/array_mediator_svc.py#L244), I see that we can receive or expect to receive more than one volume as target since we use `!= 1` instead of `== 0`

Reply:
because we didn't want to continue regularly in such an unexpected scenario (it was never supposed to happen)

Reply:
I see. It confused me to think it's a legitimate scenario.

Reply:
hyperswap does not support being a target of a fcmap, so besides its internal fcmaps, it won't have additional fcmaps as a target.